vm/src/memory/constant_pool.cpp

```diff
@@ -0,0 +1,60 @@
+#include "constant_pool.h"
+#include "byte_reader.h"
+#include "conversions.h"
+#include "strings.h"
+#include <vector>
+
+PoolConstantValue::PoolConstantValue(PoolConstantType type) : type{type} {}
+
+template <typename T>
+class PoolConstantWrappedValue : public PoolConstantValue {
+public:
+    const T value;
+    explicit PoolConstantWrappedValue(PoolConstantType type, T value) : PoolConstantValue{type}, value{std::move(value)} {}
+};
+
+std::unique_ptr<PoolConstantWrappedValue<const std::string *>> read_string(ByteReader &reader, strings_t &strings) {
+    // Read the length
+    size_t length = ntohl(reader.read<size_t>());
+
+    // Allocate the string
+    std::string str(reader.read_n<char>(length), length);
+
+    auto [str_it, _] = strings.insert(std::make_unique<std::string>(str));
```

Comment:
```suggestion
    std::string str(reader.read_n<char>(length), length);

    auto [str_it, _] = strings.insert(std::make_unique<std::string>(std::move(str)));
```

You are creating two strings, one by reading, and the other by calling `std::make_unique()`, that forward the arguments to the constructor. It is here the copy constructor.